### PR TITLE
Weight decay exclusions

### PR DESCRIPTION
**Pull request by michaellavelle**
Excluding Parameters from weight decay by default which are:

*  Freestanding ( not nested with a Module)
*  Within LayerNorm or Embedding modules

A list of parameter name overrides can be configured with TrainerConfig that allow specific Parameters to be included with weight decay

### CODE REVIEW

File: mingpt/trainer.py [hunk 1]
```diff
@@ -56,11 +57,37 @@ def save_checkpoint(self):
...
+        return list(parameters)
+
+    def no_decay_parameter_names(self, model, decay_parameter_name_overrides=None):
+
+        # define module types we always want to exclude from weight decay
+        no_decay_module_types = (torch.nn.modules.normalization.LayerNorm, torch.nn.modules.sparse.Embedding)
+
+        all_module_parameter_names = self.parameter_names_for_module_types(model, (torch.nn.Module,))
+
+        freestanding_parameter_names = [n for n, p in model.named_parameters() if not any(nd in n for nd in all_module_parameter_names)]
```
Comment (michaellavelle) [inline mingpt/trainer.py:76]: I deliberately didn't use named_parameters(recurse=False) here because we want to find *all* freestanding parameter_names,  and not just those defined at the top-level.

**Comment by karpathy** [comment]
Fixed in https://github.com/karpathy/minGPT/commit/bbbdac74fa9b2e55574d70056163ffbae42310c1

**Comment by morganmcg1** [comment]
@michaellavelle do you mind my asking why we exclude LayerNorm and Embeddings from weight decay, is there a paper/blog on this or is it more like best practice? Thanks!

**Comment by karpathy** [comment]
@morganmcg1 the purpose of L2 regularization is to "spread out" the weights in dot products, ensuring that more "independent measurements" (dimensions of the input) get used more equally, instead of any one feature dominating the computation. This only makes sense for matrix multiply layers, which the embeddings and the layernorm parameters are not. L2 regularization is covered eg in CS231n.

**Comment by morganmcg1** [comment]
@karpathy Gotcha, thank you! Some revision needed alright!
